File: algorithms/dynamic-programming/dungeon-game/java/DungeonGame.java
public class DungeonGame {

    public static int dungeonGame(int[][] grid) {
        int m = grid.length;
        if (m == 0) return 0;
        int n = grid[0].length;

        int[][] dp = new int[m][n];

        for (int i = m - 1; i >= 0; i--) {
            for (int j = n - 1; j >= 0; j--) {
                if (i == m - 1 && j == n - 1) {
                    dp[i][j] = Math.min(0, grid[i][j]);
                } else if (i == m - 1) {
                    dp[i][j] = Math.min(0, grid[i][j] + dp[i][j + 1]);
                } else if (j == n - 1) {
                    dp[i][j] = Math.min(0, grid[i][j] + dp[i + 1][j]);
                } else {
                    dp[i][j] = Math.min(0, grid[i][j] + Math.max(dp[i][j + 1], dp[i + 1][j]));
                }
            }
        }

        return Math.abs(dp[0][0]) + 1;
    }

    public static void main(String[] args) {
        int[][] grid = {
            {-2, -3, 3},
            {-5, -10, 1},
            {10, 30, -5}
        };
        System.out.println(dungeonGame(grid)); // 7
    }
}

File: algorithms/dynamic-programming/dungeon-game/kotlin/DungeonGame.kt
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

fun dungeonGame(grid: Array<IntArray>): Int {
    val m = grid.size
    if (m == 0) return 0
    val n = grid[0].size

    val dp = Array(m) { IntArray(n) }

    for (i in m - 1 downTo 0) {
        for (j in n - 1 downTo 0) {
            dp[i][j] = when {
                i == m - 1 && j == n - 1 -> min(0, grid[i][j])
                i == m - 1 -> min(0, grid[i][j] + dp[i][j + 1])
                j == n - 1 -> min(0, grid[i][j] + dp[i + 1][j])
                else -> min(0, grid[i][j] + max(dp[i][j + 1], dp[i + 1][j]))
            }
        }
    }

    return abs(dp[0][0]) + 1
}

fun main() {
    val grid = arrayOf(
        intArrayOf(-2, -3, 3),
        intArrayOf(-5, -10, 1),
        intArrayOf(10, 30, -5)
    )
    println(dungeonGame(grid)) // 7
}
